File: packages/enriched-markdown-ios/Sources/EnrichedMarkdown/Rendering/Renderers/BlankLineRenderer.swift
import UIKit

// Renders a run of consecutive blank lines emitted when preserveBlankLines is
// enabled. Each blank line in the source is drawn as one empty line, using the
// paragraph font, line height and alignment so its vertical rhythm matches the
// surrounding paragraphs. Any extra block spacing is left to the caller to
// configure.
final class BlankLineRenderer: NodeRenderer {
    private let config: MarkdownStyleConfig

    init(config: MarkdownStyleConfig) {
        self.config = config
    }

    func render(node: MarkdownASTNode, into output: NSMutableAttributedString, context: RenderContext) {
        let count = Int(node.attribute("count") ?? "0") ?? 0
        guard count > 0 else { return }

        if output.length > 0, !output.string.hasSuffix("\n") {
            output.append(ParagraphStyleHelpers.newline)
        }

        let paragraph = config.paragraph
        let font = paragraph.font ?? UIFont.preferredFont(forTextStyle: .body)

        let start = output.length
        output.append(NSAttributedString(string: String(repeating: "\n", count: count), attributes: [.font: font]))
        let range = NSRange(location: start, length: output.length - start)

        if let lineHeight = paragraph.lineHeight {
            ParagraphStyleHelpers.applyBlockLineHeight(to: output, range: range, lineHeight: lineHeight)
        }
        if let alignment = paragraph.textAlignment {
            ParagraphStyleHelpers.applyTextAlignment(to: output, range: range, alignment: alignment)
        }
    }
}

File: packages/android-enriched-markdown/ui/src/main/java/com/swmansion/enriched/markdown/renderer/BlankLineRenderer.kt
package com.swmansion.enriched.markdown.renderer

import android.text.SpannableStringBuilder
import com.swmansion.enriched.markdown.parser.MarkdownASTNode
import com.swmansion.enriched.markdown.utils.text.span.applyLineHeightSkippingImages

// Renders a run of consecutive blank lines emitted when preserveBlankLines is
// enabled. Each blank line in the source is drawn as one empty line, using the
// paragraph line height so its vertical rhythm matches the surrounding
// paragraphs. Any extra block spacing comes from the paragraph style and is left
// to the caller to configure.
class BlankLineRenderer(
  private val config: RendererConfig,
) : NodeRenderer {
  override fun render(
    node: MarkdownASTNode,
    builder: SpannableStringBuilder,
    onLinkPress: ((String) -> Unit)?,
    onLinkLongPress: ((String) -> Unit)?,
    factory: RendererFactory,
  ) {
    val count = node.getAttribute("count")?.toIntOrNull() ?: 0
    if (count <= 0) {
      return
    }

    builder.ensureNewline()
    val start = builder.length
    builder.append("\n".repeat(count))
    applyLineHeightSkippingImages(builder, start, builder.length, config.style.paragraphStyle.lineHeight)
  }

  private fun SpannableStringBuilder.ensureNewline() {
    if (isNotEmpty() && this[length - 1] != '\n') {
      append('\n')
    }
  }
}
